Shulker spawning in end cities

Salandora opened the pull request:
Ported ShulkerSpawningInEndCities feature

Also made a fix for EndCity saving as every EndCity saved once got invalidated at a server restart because a boolean value was not saved.
This was also breaking the locate function to find already visited EndCities.

Earthcomputer commented on carpetmodSrc/carpet/commands/CommandAstral.java:
```diff
@@ -0,0 +1,120 @@
+     */
+    public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+        if (!command_enabled("commandAstral", sender)) {
+            throw new WrongUsageException("Enable commandAstral to use this command.");
```
WrongUsageException is for when the command syntax is wrong, as it will display "Usage:" prefix. Use CommandException instead.

Earthcomputer commented on carpetmodSrc/carpet/commands/CommandAstral.java:
```diff
@@ -0,0 +1,120 @@
+            throw new WrongUsageException("Enable commandAstral to use this command.");
+        }
+        if (!(sender instanceof EntityPlayerMP)) {
+            throw new WrongUsageException("Only players can use this command.");
```
Same here

Earthcomputer commented on carpetmodSrc/carpet/commands/CommandAstral.java:
```diff
@@ -0,0 +1,120 @@
+                String playerName = args[1];
+                EntityPlayerMP kick = server.getPlayerList().getPlayerByUsername(playerName);
+                if (!(kick instanceof EntityPlayerMPFake)) {
+                    throw new WrongUsageException("/astral kick can only be used on astral or fake players");
```
And here

Earthcomputer commented on carpetmodSrc/carpet/commands/CommandAstral.java:
```diff
@@ -0,0 +1,120 @@
+        if(args.length > 1)
+            System.out.println("testing :" + args[0]+":1");
+        if (args.length == 1) {
+            return getListOfStringsMatchingLastWord(args, new String[]{"leftClick", "rightClick", "leftAndRight", "kick"});
```
This would be cleaner with varargs

Earthcomputer commented on carpetmodSrc/carpet/patches/EntityPlayerMPFake.java:
```diff
@@ -1,26 +1,42 @@
+
+import net.minecraft.world.GameType;
+
+public class EntityPlayerMPFake extends EntityPlayerMP {
```
A couple of braces inconsistent with the rest of the file

Earthcomputer commented on carpetmodSrc/carpet/patches/EntityPlayerMPFake.java:
```diff
@@ -56,38 +71,87 @@ public static EntityPlayerMPFake createShadow(MinecraftServer server, EntityPlay
     }
 
-    private EntityPlayerMPFake(MinecraftServer server, WorldServer worldIn, GameProfile profile, PlayerInteractionManager interactionManagerIn)
-    {
```
Braces again

Earthcomputer commented on carpetmodSrc/carpet/patches/EntityPlayerMPFake.java:
```diff
@@ -56,38 +71,87 @@ public static EntityPlayerMPFake createShadow(MinecraftServer server, EntityPlay
+        return astral;
+    }
+
+    private static GameProfile profileCasher(GameProfile gameprofile) {
```
Usually not picky with spelling but this one can be confusing, should be "profileCacher"

Earthcomputer commented:
Can't request changes on my own PR :(

Earthcomputer commented:
Looks good now

gnembon commented:
wtf is astral. I removed it cause it is part of /player commands as /player x shadow

gnembon commented on carpetmodSrc/carpet/patches/EntityPlayerMPFake.java:
```diff
@@ -56,32 +74,84 @@ public static EntityPlayerMPFake createShadow(MinecraftServer server, EntityPlay
+        return astral;
+    }
+
+    private static GameProfile profileCacher(GameProfile gameprofile)
```
its not a cache. btw this entire block is just to do
```
if (gameprofile.getProperties().containsKey("textures"))
        {
            gameprofile = TileEntitySkull.updateGameProfile(gameprofile);
        }
```
inside the constructor

gnembon commented on patches/net/minecraft/entity/Entity.java.patch:
```diff
@@ -142,5 +142,9 @@
 +    
 +    public void postLoad()
 +    {
++    }
```
doesn't need to be in Entity. It applies only to players.

gnembon commented on patches/net/minecraft/server/management/PlayerList.java.patch:
```diff
@@ -21,7 +21,16 @@
+             nbttagcompound1 = this.field_72412_k.func_75752_b(p_72380_1_);
+         }
+ 
++        p_72380_1_.postReadFromNBT();
```
here if checked if p_7... instanceof EPMPF than all of the extra fields in base classes are not needed.

Earthcomputer commented:
Astral is an attempt to address the current reality that fake players are being used in survival -- it's supposed to be a less cheaty version of it. However, I don't think it's a complete solution so more discussion is needed, which I will create an issue for. The fakeplayer fixes however there's really no question that they can go in, and taking your extra review into account I think I'll just commit them directly as with other bugfixes. All that said, there is no reason left for this PR to exist, so I'll close it.

Salandora committed:
Ported ShulkerSpawningInEndCities feature

Also fixed EndCity saving as every EndCity saved once got invalidated at a server restart because a boolean value was not saved.
This was also breaking the locate function to find already visited citys

Salandora committed:
Merge branch 'master' of github.com:gnembon/carpetmod into shulkerSpawningInEndCities

# Conflicts:
#	src/main/java/carpet/CarpetSettings.java

Salandora committed:
Updated min and max group size to reflect the original values.

gnembon commented on src/main/java/carpet/CarpetSettings.java:
```diff
@@ -137,7 +138,7 @@ private static void set_defaults()
   //                              .extraInfo("Also prevents rotations upon placement of dispensers and furnaces","when placed into a world by commands"),
   /////rule("optimizedTNT",          "tnt", "TNT causes less lag when exploding in the same spot and in liquids"),
   /////rule("huskSpawningInTemples", "experimental feature", "Only husks spawn in desert temples"),
-  /////rule("shulkerSpawningInEndCities", "feature experimental", "Shulkers will respawn in end cities"),
```
this was marked with ///// meaning - do not port...
I heard there were performance issues with this one.

Salandora replied:
Ohh okay.
Performance problems, hmm, I'll have a close look at it as I'm using it anyway.

Earthcomputer replied:
The performance issues were likely due to the fact that `recreateStructures` was called in the end, even when the rule was disabled. I don't know whether this would affect 1.13 due to the world gen rewrite.

Salandora commented:
As @gnembon said in his line comment it was not supposed to be ported so I'm closing it.

Salandora commented:
Reopen because of recent talk in discord

gnembon commented on patches/net/minecraft/world/gen/feature/structure/EndCityStructure.java.patch:
```diff
@@ -0,0 +1,67 @@
+         return 9;
+     }
+ 
++    public List<Biome.SpawnListEntry> getSpawnList()
```
@Override missing

gnembon commented on patches/net/minecraft/world/gen/feature/structure/EndCityStructure.java.patch:
```diff
@@ -0,0 +1,67 @@
++                {
++                    // Called the tag Valid because it is called Valid in the VillageStructure too
++                    // and it is used to invalidate the Structure
++                    tagCompound.setBoolean("Valid", this.isSizeable);
```
I will be frank - not fond of this one. Unless it fixes some critical issue with vanilla, then I would abstain from it, even if it is a valid issue. Especially the comment doesn't convince me - villages are the only temporary structures - the rest is permanent, and extra tags may not be necessary.

Salandora replied:
Well In 1.13 the end city saving is bugged.
Just try this: 
Enter the end and type "/locate EndCity" in the console and teleport there.
Then teleport back to end spawn island. Save and restart the server and try "/locate EndCity" again.
It will not find the End City of the first run any more.
While debugging I noticed that the End City, once generated and saved to file get's invalidated on reload.

But if you like I can remove it.

DeadlyMC replied:
Is it only with the EndCity though?
It maybe considered intended cause most people don't want the same EndCity to appear everytime u run /locate.

Salandora replied:
It is also with Woodland Mansions.
But you might be right, it could be intended.

Salandora committed:
Added missing @Override

gnembon committed:
removed NBT modifications
